File: scripts/run_loop.py
#!/usr/bin/env python3
"""Trigger-accuracy eval loop for the SKILL.md `description`.

A skill only helps if its `description` fires on the right prompts and stays
quiet on the rest. This measures that against a labelled eval set so you can
iterate the description deliberately instead of guessing.

The fire/no-fire decision ultimately belongs to the model that reads the
description at runtime, which a script can't invoke. So this offers two paths:

  * Offline baseline (default, --heuristic): a cheap token-overlap classifier
    using trigger words mined from the description. Lets the whole loop run
    with no model in the harness — good for catching obvious gaps fast.

  * Real judgments (--predictions preds.json): score predictions produced by
    Claude (or any judge). `preds.json` is {"prompt": "fire"|"no_fire", ...}.
    This is the accurate path; the heuristic is only a stand-in.

Reports precision / recall / F1 / trigger-rate, lists false positives and
false negatives, and suggests words to add to (or reconsider in) the
description — the "improvement" half of the loop.

Usage:
    python scripts/run_loop.py --skill SKILL.md --eval evals/trigger_evalset.json
    python scripts/run_loop.py --eval evals/trigger_evalset.json \
        --predictions /tmp/preds.json --json
    python scripts/run_loop.py --make-eval > evals/trigger_evalset.json
"""

import argparse
import json
import os
import re
import sys

TOKEN_RE = re.compile(r"[a-z0-9][a-z0-9_+-]*")

STOPWORDS = {
    "the", "a", "an", "to", "of", "in", "on", "for", "and", "or", "is", "are",
    "this", "that", "these", "those", "with", "your", "you", "i", "it", "if",
    "do", "does", "can", "could", "would", "should", "any", "my", "me", "we",
    "be", "been", "as", "at", "by", "from", "into", "out", "over", "use", "uses",
    "so", "but", "not", "no", "yes", "what", "which", "how", "they", "them",
    "their", "its", "than", "then", "when", "where", "who", "will", "shall",
    "there", "here", "about", "before", "after", "more", "most", "every",
    "each", "all", "some", "one", "two", "e", "g",
}

# A small curated lexicon of strong trigger words for this skill, always
# considered part of the description's vocabulary even if phrasing changes.
CORE_TRIGGERS = {
    "review", "diff", "branch", "pr", "pull", "request", "changes", "changeset",
    "blast", "radius", "impact", "callers", "affected", "staged", "modified",
    "patch", "risky", "audit",
}


def tokens(text):
    return [t for t in TOKEN_RE.findall(text.lower())
            if len(t) > 1 and t not in STOPWORDS]


def read_description(skill_path):
    """Extract the YAML frontmatter `description` from a SKILL.md."""
    with open(skill_path, "r", encoding="utf-8") as fh:
        text = fh.read()
    m = re.search(r"^---\s*\n(.*?)\n---\s*\n", text, re.DOTALL)
    block = m.group(1) if m else text
    dm = re.search(r"(?ms)^description:\s*(.*?)(?=^\w[\w-]*:\s|\Z)", block)
    if not dm:
        return ""
    raw = dm.group(1)
    # strip YAML block scalar markers and indentation
    raw = re.sub(r"^[>|][+-]?\s*", "", raw.strip())
    return " ".join(line.strip() for line in raw.splitlines())


def load_eval(path):
    with open(path, "r", encoding="utf-8") as fh:
        data = json.load(fh)
    cases = data["cases"] if isinstance(data, dict) else data
    for c in cases:
        if c.get("label") not in ("should_fire", "should_not_fire"):
            raise ValueError("bad label in eval case: %r" % c)
    return cases


def heuristic_predictions(cases, description, threshold):
    vocab = set(tokens(description)) | CORE_TRIGGERS
    preds = {}
    scores = {}
    for c in cases:
        toks = set(tokens(c["prompt"]))
        hits = toks & vocab
        score = len(hits)
        scores[c["prompt"]] = sorted(hits)
        preds[c["prompt"]] = "fire" if score >= threshold else "no_fire"
    return preds, scores, vocab


def score(cases, preds):
    tp = fp = tn = fn = 0
    false_pos, false_neg = [], []
    for c in cases:
        want_fire = c["label"] == "should_fire"
        got = preds.get(c["prompt"], "no_fire")
        got_fire = got == "fire"
        if want_fire and got_fire:
            tp += 1
        elif want_fire and not got_fire:
            fn += 1
            false_neg.append(c["prompt"])
        elif not want_fire and got_fire:
            fp += 1
            false_pos.append(c["prompt"])
        else:
            tn += 1
    precision = tp / (tp + fp) if (tp + fp) else 1.0
    recall = tp / (tp + fn) if (tp + fn) else 1.0
    f1 = (2 * precision * recall / (precision + recall)) if (precision + recall) else 0.0
    trigger_rate = (tp + fp) / len(cases) if cases else 0.0
    return {
        "tp": tp, "fp": fp, "tn": tn, "fn": fn,
        "precision": round(precision, 3),
        "recall": round(recall, 3),
        "f1": round(f1, 3),
        "trigger_rate": round(trigger_rate, 3),
        "false_positives": false_pos,
        "false_negatives": false_neg,
    }


def suggestions(cases, preds, description):
    """Mine words to add (from FNs) / reconsider (from FPs)."""
    desc_vocab = set(tokens(description)) | CORE_TRIGGERS
    fn_words, fp_words = {}, {}
    for c in cases:
        want_fire = c["label"] == "should_fire"
        got_fire = preds.get(c["prompt"], "no_fire") == "fire"
        if want_fire and not got_fire:
            for t in set(tokens(c["prompt"])) - desc_vocab:
                fn_words[t] = fn_words.get(t, 0) + 1
        if not want_fire and got_fire:
            for t in set(tokens(c["prompt"])) & desc_vocab:
                fp_words[t] = fp_words.get(t, 0) + 1
    add = [w for w, _ in sorted(fn_words.items(), key=lambda kv: -kv[1])][:8]
    reconsider = [w for w, _ in sorted(fp_words.items(), key=lambda kv: -kv[1])][:8]
    return {"consider_adding": add, "ambiguous_in_description": reconsider}


STARTER_EVAL = {
    "description": "Starter eval set for SKILL.md trigger tuning. Replace/extend with prompts you actually see.",
    "cases": [
        {"prompt": "Review this branch against main", "label": "should_fire"},
        {"prompt": "What's the blast radius of this change?", "label": "should_fire"},
        {"prompt": "Write a function that reverses a linked list", "label": "should_not_fire"},
        {"prompt": "What's the capital of France?", "label": "should_not_fire"}
    ],
}


def main():
    p = argparse.ArgumentParser(description="SKILL.md trigger-accuracy eval loop")
    here = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    p.add_argument("--skill", default=os.path.join(here, "SKILL.md"))
    p.add_argument("--eval", default=os.path.join(here, "evals", "trigger_evalset.json"))
    p.add_argument("--predictions", help="JSON {prompt: 'fire'|'no_fire'} from a real judge")
    p.add_argument("--heuristic", action="store_true",
                   help="force offline heuristic even if --predictions given")
    p.add_argument("--threshold", type=int, default=1,
                   help="min trigger-word hits for heuristic fire (default 1)")
    p.add_argument("--json", action="store_true", help="emit JSON instead of text")
    p.add_argument("--make-eval", action="store_true", help="print a starter eval set and exit")
    args = p.parse_args()

    if args.make_eval:
        json.dump(STARTER_EVAL, sys.stdout, indent=2)
        sys.stdout.write("\n")
        return

    cases = load_eval(args.eval)
    description = read_description(args.skill)

    if args.predictions and not args.heuristic:
        with open(args.predictions, "r", encoding="utf-8") as fh:
            preds = json.load(fh)
        mode = "predictions"
        hit_words = {}
    else:
        preds, hit_words, _ = heuristic_predictions(cases, description, args.threshold)
        mode = "heuristic"

    metrics = score(cases, preds)
    sugg = suggestions(cases, preds, description)

    report = {
        "mode": mode,
        "skill": os.path.relpath(args.skill),
        "eval_set": os.path.relpath(args.eval),
        "n_cases": len(cases),
        "metrics": metrics,
        "suggestions": sugg,
    }

    if args.json:
        json.dump(report, sys.stdout, indent=2)
        sys.stdout.write("\n")
        return

    m = metrics
    print("trigger eval (%s) over %d cases" % (mode, len(cases)))
    print("  precision=%.3f recall=%.3f f1=%.3f trigger_rate=%.3f"
          % (m["precision"], m["recall"], m["f1"], m["trigger_rate"]))
    print("  tp=%d fp=%d tn=%d fn=%d" % (m["tp"], m["fp"], m["tn"], m["fn"]))
    if m["false_negatives"]:
        print("  MISSED (should fire but didn't):")
        for q in m["false_negatives"]:
            print("    - %s" % q)
    if m["false_positives"]:
        print("  OVER-FIRED (fired but shouldn't):")
        for q in m["false_positives"]:
            print("    - %s" % q)
    if sugg["consider_adding"]:
        print("  consider adding to description: %s" % ", ".join(sugg["consider_adding"]))
    if sugg["ambiguous_in_description"]:
        print("  ambiguous trigger words (cause over-fire): %s"
              % ", ".join(sugg["ambiguous_in_description"]))


if __name__ == "__main__":
    main()

File: js/run_loop.mjs
#!/usr/bin/env node
// Trigger-accuracy eval loop — Node port of scripts/run_loop.py.

import { readFileSync } from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

const HERE = path.dirname(path.dirname(new URL(import.meta.url).pathname));
const TOKEN_RE = /[a-z0-9][a-z0-9_+-]*/g;

const STOPWORDS = new Set(("the a an to of in on for and or is are this that these those with your you i it if " +
  "do does can could would should any my me we be been as at by from into out over use uses so but not no yes " +
  "what which how they them their its than then when where who will shall there here about before after more " +
  "most every each all some one two e g").split(" "));

const CORE_TRIGGERS = new Set(["review", "diff", "branch", "pr", "pull", "request", "changes", "changeset",
  "blast", "radius", "impact", "callers", "affected", "staged", "modified", "patch", "risky", "audit"]);

function tokens(text) {
  const out = [];
  for (const m of text.toLowerCase().matchAll(TOKEN_RE)) {
    const t = m[0];
    if (t.length > 1 && !STOPWORDS.has(t)) out.push(t);
  }
  return out;
}

function readDescription(skillPath) {
  const text = readFileSync(skillPath, "utf8");
  const m = text.match(/^---\s*\n([\s\S]*?)\n---\s*\n/);
  const block = m ? m[1] : text;
  const lines = block.split("\n");
  const i = lines.findIndex((l) => /^description:/.test(l));
  if (i < 0) return "";
  const parts = [lines[i].replace(/^description:\s*/, "")];
  for (let j = i + 1; j < lines.length; j++) {
    if (/^\w[\w-]*:\s/.test(lines[j])) break; // next top-level key
    parts.push(lines[j]);
  }
  const raw = parts.join("\n").trim().replace(/^[>|][+-]?\s*/, "");
  return raw.split("\n").map((l) => l.trim()).join(" ");
}

function loadEval(p) {
  const data = JSON.parse(readFileSync(p, "utf8"));
  const cases = Array.isArray(data) ? data : data.cases;
  for (const c of cases) if (c.label !== "should_fire" && c.label !== "should_not_fire") throw new Error("bad label: " + JSON.stringify(c));
  return cases;
}

function heuristicPredictions(cases, description, threshold) {
  const vocab = new Set([...tokens(description), ...CORE_TRIGGERS]);
  const preds = {};
  for (const c of cases) {
    const toks = new Set(tokens(c.prompt));
    let hits = 0;
    for (const t of toks) if (vocab.has(t)) hits++;
    preds[c.prompt] = hits >= threshold ? "fire" : "no_fire";
  }
  return preds;
}

function score(cases, preds) {
  let tp = 0, fp = 0, tn = 0, fn = 0;
  const falsePos = [], falseNeg = [];
  for (const c of cases) {
    const wantFire = c.label === "should_fire";
    const gotFire = (preds[c.prompt] || "no_fire") === "fire";
    if (wantFire && gotFire) tp++;
    else if (wantFire && !gotFire) { fn++; falseNeg.push(c.prompt); }
    else if (!wantFire && gotFire) { fp++; falsePos.push(c.prompt); }
    else tn++;
  }
  const precision = (tp + fp) ? tp / (tp + fp) : 1.0;
  const recall = (tp + fn) ? tp / (tp + fn) : 1.0;
  const f1 = (precision + recall) ? (2 * precision * recall) / (precision + recall) : 0.0;
  const triggerRate = cases.length ? (tp + fp) / cases.length : 0.0;
  const r3 = (x) => Math.round(x * 1000) / 1000;
  return { tp, fp, tn, fn, precision: r3(precision), recall: r3(recall), f1: r3(f1),
    trigger_rate: r3(triggerRate), false_positives: falsePos, false_negatives: falseNeg };
}

function suggestions(cases, preds, description) {
  const descVocab = new Set([...tokens(description), ...CORE_TRIGGERS]);
  const fnWords = {}, fpWords = {};
  for (const c of cases) {
    const wantFire = c.label === "should_fire";
    const gotFire = (preds[c.prompt] || "no_fire") === "fire";
    const toks = new Set(tokens(c.prompt));
    if (wantFire && !gotFire) for (const t of toks) if (!descVocab.has(t)) fnWords[t] = (fnWords[t] || 0) + 1;
    if (!wantFire && gotFire) for (const t of toks) if (descVocab.has(t)) fpWords[t] = (fpWords[t] || 0) + 1;
  }
  const top = (o) => Object.entries(o).sort((a, b) => b[1] - a[1]).map((e) => e[0]).slice(0, 8);
  return { consider_adding: top(fnWords), ambiguous_in_description: top(fpWords) };
}

const STARTER_EVAL = {
  description: "Starter eval set for SKILL.md trigger tuning. Replace/extend with prompts you actually see.",
  cases: [
    { prompt: "Review this branch against main", label: "should_fire" },
    { prompt: "What's the blast radius of this change?", label: "should_fire" },
    { prompt: "Write a function that reverses a linked list", label: "should_not_fire" },
    { prompt: "What's the capital of France?", label: "should_not_fire" },
  ],
};

function parseArgs(argv) {
  const a = { skill: path.join(HERE, "SKILL.md"), eval: path.join(HERE, "evals", "trigger_evalset.json"),
    predictions: null, heuristic: false, threshold: 1, json: false, makeEval: false };
  for (let i = 0; i < argv.length; i++) {
    const x = argv[i];
    if (x === "--skill") a.skill = argv[++i];
    else if (x === "--eval") a.eval = argv[++i];
    else if (x === "--predictions") a.predictions = argv[++i];
    else if (x === "--heuristic") a.heuristic = true;
    else if (x === "--threshold") a.threshold = +argv[++i];
    else if (x === "--json") a.json = true;
    else if (x === "--make-eval") a.makeEval = true;
  }
  return a;
}

function main() {
  const a = parseArgs(process.argv.slice(2));
  if (a.makeEval) { process.stdout.write(JSON.stringify(STARTER_EVAL, null, 2) + "\n"); return; }

  const cases = loadEval(a.eval);
  const description = readDescription(a.skill);

  let preds, mode;
  if (a.predictions && !a.heuristic) { preds = JSON.parse(readFileSync(a.predictions, "utf8")); mode = "predictions"; }
  else { preds = heuristicPredictions(cases, description, a.threshold); mode = "heuristic"; }

  const metrics = score(cases, preds);
  const sugg = suggestions(cases, preds, description);
  const report = { mode, skill: path.relative(".", a.skill), eval_set: path.relative(".", a.eval),
    n_cases: cases.length, metrics, suggestions: sugg };

  if (a.json) { process.stdout.write(JSON.stringify(report, null, 2) + "\n"); return; }
  const m = metrics;
  console.log(`trigger eval (${mode}) over ${cases.length} cases`);
  console.log(`  precision=${m.precision.toFixed(3)} recall=${m.recall.toFixed(3)} f1=${m.f1.toFixed(3)} trigger_rate=${m.trigger_rate.toFixed(3)}`);
  console.log(`  tp=${m.tp} fp=${m.fp} tn=${m.tn} fn=${m.fn}`);
  if (m.false_negatives.length) { console.log("  MISSED (should fire but didn't):"); for (const q of m.false_negatives) console.log("    - " + q); }
  if (m.false_positives.length) { console.log("  OVER-FIRED (fired but shouldn't):"); for (const q of m.false_positives) console.log("    - " + q); }
  if (sugg.consider_adding.length) console.log("  consider adding to description: " + sugg.consider_adding.join(", "));
  if (sugg.ambiguous_in_description.length) console.log("  ambiguous trigger words (cause over-fire): " + sugg.ambiguous_in_description.join(", "));
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) main();
